### PR TITLE
Standalone mcode client

### PR DESCRIPTION
**Pull request by Dtphelan1**
Standalone mCODE Client 

# Summary
Creates a mCODE client capable of extracting mCODE data using all MEF extractors, and adds a CLI that uses this client - along with other utilities - to extract patient data based on the supplied configuration file, and to write that data to disk. 

## New behavior
- Moves over the following Application-utilities from base-icare-client to this repo: 
  - RunInstanceLogger
  - emailNotifications
- Creates a mCODEClient that registers all mCODE extractors. 
- Creates a cliApplication that uses the mCODEClient as well as a few other application-utilities to extract mCODE data and write it to disk

## Code changes
Most everything changed is captured in the above new behavior, with exception of: 
- updates to the .gitignore to avoid config files and output directory info
- *IMPT*: Updating of the authentication information now happens within the `init` function, a member-function that is common to all `client` classes. Housing the initialitzation of our class-instances in this `init` function allows us to make asynchronous calls in this location, which is necessary for getting authentication tokens. 

# Testing guidance
- Ensure tests pass. 
- Ensure you understand and are comfortable with the breakdown of application utilities, Client code, and what has been separated from/left in the base-icare-client repo.

### CODE REVIEW

**Review by jafeltra — COMMENTED** [review]
This looks really great!! I really think the division of the different pieces of code, and I think that having an mCODE client will simplify some of the questions we were having around the ICARE client. I just had a few things I noted in line. Some of these are just questions for my own understanding and don't necessarily require any changes.

File: src/cli/mcodeExtraction.js [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+      totalExtractionErrors[index].push(fatalErr);
+      logger.error(`Fatal error extracting data: ${fatalErr.message}`);
+      logger.debug(fatalErr.stack);
+    }
+  }
+  return { extractedData, successfulExtraction, totalExtractionErrors };
+}
+
+module.exports = {
+  getResourceCountInBundle,
```
Comment (danlee1025) [inline src/cli/mcodeExtraction.js:33]: [DL1] Is there a particular why this is also exported here?

File: config/mcode-csv-config.example.json [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+        "clinicalSiteID": "example-site-id"
+      }
+    },
+    {
+      "label": "treatmentPlanChange",
+      "type": "CSVTreatmentPlanChangeExtractor",
+      "constructorArgs": {
+        "filePath": "./data/treatment-plan-change-information.csv"
+      }
+    }
```
[the remark this answers is not in the document]
Comment (Dtphelan1) [reply config/mcode-csv-config.example.json:49]: I don't have strong opinions here. Happy to add them and a staging one though. 😄

File: test/cli/emailNotifications.test.js [hunk 3]
```diff
@@ -0,0 +1,119 @@
+// Not entirely sure why nodemailer can't be imported directly and must be imported this way
```
[the remark this answers is not in the document]
Comment (Dtphelan1) [reply test/cli/emailNotifications.test.js:1]: Addressed!

File: src/cli/app.js [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+    if (notificationInfo) {
+      const notificationErrors = zipErrors(totalExtractionErrors);
+      await sendEmailNotification(notificationInfo, notificationErrors, debug);
+    }
+    // A run is successful and should be logged when both extraction finishes without fatal errors
+    // and messages are posted without fatal errors
+    if (!allEntries && effectiveFromDate) {
+      const successCondition = successfulExtraction;
+      if (successCondition) {
+        logger.info('Logging successful run information to records');
```
[the remark this answers is not in the document]
Comment (Dtphelan1) [reply src/cli/app.js:102]: Removed!

File: config/mcode-csv-config.example.json (hunk 2 above)
[the remark this answers is not in the document]
Comment (Dtphelan1) [reply config/mcode-csv-config.example.json:77]: Added!

File: test/helpers/dateUtils.test.js [hunk 5]
```diff
@@ -11,7 +11,7 @@ test('formatDate does not reformat invalid date', () => {
 });
 
 test('formatDateTime reformats date with a time if provided', () => {
-  const currentTimeZone = moment().format('Z');
+  const currentTimeZone = moment('04/12/19').format('Z');
```
[the remark this answers is not in the document]
Comment (Dtphelan1) [reply test/helpers/dateUtils.test.js]: 😆

File: test/helpers/fhirUtils.test.js [hunk 6]
```diff
@@ -62,3 +62,9 @@ test('allResourcesInBundle', () => {
   expect(allResourcesInBundle(bundleWithMultipleEntries))
     .toEqual(bundleWithMultipleEntries.entry.map((e) => e.resource));
 });
+
+
+test('getResourceCountInBundle', () => {
```
[the remark this answers is not in the document]
Comment (Dtphelan1) [reply test/helpers/fhirUtils.test.js:67]: Addressed in an additional task 👍

File: src/cli/mcodeExtraction.js (hunk 1 above)
Comment (Dtphelan1) [reply src/cli/mcodeExtraction.js]: lol nope

File: test/cli/fixtures/test-bundle.json [hunk 7]
```diff
@@ -0,0 +1,106 @@
+{
+  "resourceType": "Bundle",
+  "id": "32e1d04d-73f4-4dbc-802b-13dce4fdc87b",
+  "type": "message",
+  "timestamp": "2020-06-24T11:17:03-04:00",
+  "entry": [
+    {
+      "fullUrl": "urn:uuid:0408e9ec-cf22-425b-bf8a-683ac4b7335e",
+      "resource": {
+        "resourceType": "MessageHeader",
```
[the remark this answers is not in the document]
Comment (Dtphelan1) [reply test/cli/fixtures/test-bundle.json]: That's a great criticism. Taken care of 👍

File: src/helpers/fhirUtils.js [hunk 8]
```diff
@@ -19,6 +19,26 @@ const quantityCodeToUnitLookup = {
   g: 'g',
 };
 
+function getResourceCountInBundle(messageBundle) {
```
[the remark this answers is not in the document]
Comment (Dtphelan1) [reply src/helpers/fhirUtils.js:22]: happy to! lmk if you're around sometime today! 👍

**Comment by Dtphelan1** [comment]
Dan re: your last comment, there is a task in the sprint right now that we'll use to update README's appropriately for the base-icare and MEF repo's, moving mcode-only elements from the base-icare into the MEF. We've updated that description to ensure that we also update the README to discuss the CLI/Client information.

File: config/mcode-csv-config.example.json [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+      "type": "CSVTreatmentPlanChangeExtractor",
+      "constructorArgs": {
+        "filePath": "./data/treatment-plan-change-information.csv"
+      }
+    },
+    {
+      "label": "staging",
+      "type": "CSVStagingExtractor",
+      "constructorArgs": {
+        "filePath": "./test/sample-client-data/staging-information.csv"
```
[the remark this answers is not in the document]
Comment (Dtphelan1) [reply config/mcode-csv-config.example.json]: Addressed.

File: src/helpers/fhirUtils.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (Dtphelan1) [reply src/helpers/fhirUtils.js:23]: We've discussed this functionality to a productive extent, and we discovered an area where this function could be improved! Thanks for chatting Jules! 😄

File: src/cli/app.js [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+    }
+
+    // Finally, save the data to disk
+    const outputPath = './output';
+    if (!fs.existsSync(outputPath)) {
+      logger.info(`Creating directory ${outputPath}`);
+      fs.mkdirSync(outputPath);
+    }
+    const outputFile = path.join(outputPath, 'mcode-extraction.json');
+    fs.writeFileSync(outputFile, JSON.stringify(extractedData), 'utf8');
```
[the remark this answers is not in the document]
Comment (Dtphelan1) [reply src/cli/app.js:114]: This should be addressed with a recent commit 😄

**Review by jafeltra — APPROVED** [review]
All the things I noticed are resolved! This looks great!

**Review by danlee1025 — APPROVED** [review]

**Pull request merged**